core/src/main/java/org/springframework/security/authorization/AuthorizationDecision.java

```diff
@@ -18,10 +18,21 @@
 
 /**
  * @author Rob Winch
+ * @author Evgeniy Cheban
  * @since 5.0
  */
 public class AuthorizationDecision {
```

Comment:
A bit more thinking out loud, here. I wonder if `AuthorizationDecision` would benefit from a collection of reasons. Similar to `OAuth2TokenValidatorResult`, it's nice to be able to see why validation succeeded or failed.

The reason I'm thinking about this is that `AuthorizedEvent` and `AuthorizedFailureEvent` are both capable, through the config attributes, of saying why it succeeded or failed. It might be reasonable to add the `AuthorizationDecision` to the event.

Reply:
Agreed, but instead of adding the `AuthorizationDecision` to the event I wonder if it would be better to create a `ConfigAttribute` collection from a collection of reasons to create the event, so that user won't notice any changes after the update. What do you think?

Reply:
The `AuthorizationDecision` can also be added to the event and the collection of `ConfigAttribute` can be marked as deprecated.

Reply:
Yes, I like this idea. For now, let's try and leave the events unchanged, and then those things that are firing the events can produce a collection of `ConfigAttribute`s.

If that works well, you might consider re-consolidating the logging and event-throwing logic back into `AbstractSecurityInterceptor`. One reason this would be nice is it would prevent having to have the `publishAuthorizationSuccess` configuration in multiple places.